From 791ba8cfc75494e7a937b55a266716b84f0ca36a Mon Sep 17 00:00:00 2001
From: Julien Marrec <julien.marrec@gmail.com>
Date: Tue, 16 Jun 2026 21:56:56 +0200
Subject: [PATCH] Allow windows arm64-mswin and fix MACHINE/size/atomic
 detection

Backport of https://github.com/ruby/ruby/pull/12367 plus arm64 build fixes
backported from Ruby 3.4.1:

- win32/setup.mak: detect _M_ARM64 so MACHINE=arm64 (not x86 fallback),
  fixing sitearch=arm64-mswin64_140 and config.h placement
- win32/Makefile.sub: gate SIZEOF_VOIDP/SIZE_T/PTRDIFF_T/INTPTR_T/UINTPTR_T,
  ssize_t and PRI_PTR_PREFIX on TARGET_OS==mswin64 instead of ARCH==x64
- include/ruby/atomic.h: use _WIN64 instead of _WIN32&&_M_AMD64 for 64-bit
  Interlocked ops so arm64 size_t atomics don't fall to the broken #else path
---
 coroutine/arm64/Context.asm    | 81 ++++++++++++++++++++++++++++++++++
 coroutine/arm64/Context.h      |  8 +++-
 include/ruby/atomic.h          | 12 ++---
 include/ruby/internal/memory.h | 11 ++++-
 vm_dump.c                      |  8 ++++
 win32/Makefile.sub             | 12 +++--
 win32/setup.mak                |  6 ++-
 7 files changed, 126 insertions(+), 12 deletions(-)
 create mode 100644 coroutine/arm64/Context.asm

diff --git a/coroutine/arm64/Context.asm b/coroutine/arm64/Context.asm
new file mode 100644
index 00000000..866fa628
--- /dev/null
+++ b/coroutine/arm64/Context.asm
@@ -0,0 +1,81 @@
+	TTL	coroutine/arm64/Context.asm
+
+	AREA	|.drectve|, DRECTVE
+
+	EXPORT	|coroutine_transfer|
+
+	AREA	|.text$mn|, CODE, ARM64
+
+;; Add more space for certain TEB values on each stack
+TEB_OFFSET EQU 0x20
+
+;; Incomplete implementation
+coroutine_transfer PROC
+	; Make space on the stack for caller registers
+	sub sp, sp, 0xa0 + TEB_OFFSET
+
+	; Save caller registers
+	stp d8, d9, [sp, 0x00 + TEB_OFFSET]
+	stp d10, d11, [sp, 0x10 + TEB_OFFSET]
+	stp d12, d13, [sp, 0x20 + TEB_OFFSET]
+	stp d14, d15, [sp, 0x30 + TEB_OFFSET]
+	stp x19, x20, [sp, 0x40 + TEB_OFFSET]
+	stp x21, x22, [sp, 0x50 + TEB_OFFSET]
+	stp x23, x24, [sp, 0x60 + TEB_OFFSET]
+	stp x25, x26, [sp, 0x70 + TEB_OFFSET]
+	stp x27, x28, [sp, 0x80 + TEB_OFFSET]
+	stp x29, x30, [sp, 0x90 + TEB_OFFSET]
+
+	;; Save certain values from Thread Environment Block (TEB) x18
+	;; points to the TEB on Windows
+	;; Read TeStackBase and TeStackLimit at ksarm64.h from TEB
+	ldp  x5,  x6,  [x18, #0x08]
+	;; Save them
+	stp  x5,  x6,  [sp, #0x00]
+	;; Read TeDeallocationStack at ksarm64.h from TEB
+	ldr  x5, [x18, #0x1478]
+	;; Read TeFiberData at ksarm64.h from TEB
+	ldr  x6, [x18, #0x20]
+	;; Save current fiber data and deallocation stack
+	stp  x5,  x6,  [sp, #0x10]
+
+	; Save stack pointer to x0 (first argument)
+	mov x2, sp
+	str x2, [x0, 0]
+
+	; Load stack pointer from x1 (second argument)
+	ldr x3, [x1, 0]
+	mov sp, x3
+
+	;; Restore stack base and limit
+	ldp  x5,  x6,  [sp, #0x00]
+	;; Write TeStackBase and TeStackLimit at ksarm64.h to TEB
+	stp  x5,  x6,  [x18, #0x08]
+	;; Restore fiber data and deallocation stack
+	ldp  x5,  x6,  [sp, #0x10]
+	;; Write TeDeallocationStack at ksarm64.h to TEB
+	str  x5, [x18, #0x1478]
+	;; Write TeFiberData at ksarm64.h to TEB
+	str  x6, [x18, #0x20]
+
+	; Restore caller registers
+	ldp d8, d9, [sp, 0x00 + TEB_OFFSET]
+	ldp d10, d11, [sp, 0x10 + TEB_OFFSET]
+	ldp d12, d13, [sp, 0x20 + TEB_OFFSET]
+	ldp d14, d15, [sp, 0x30 + TEB_OFFSET]
+	ldp x19, x20, [sp, 0x40 + TEB_OFFSET]
+	ldp x21, x22, [sp, 0x50 + TEB_OFFSET]
+	ldp x23, x24, [sp, 0x60 + TEB_OFFSET]
+	ldp x25, x26, [sp, 0x70 + TEB_OFFSET]
+	ldp x27, x28, [sp, 0x80 + TEB_OFFSET]
+	ldp x29, x30, [sp, 0x90 + TEB_OFFSET]
+
+	; Pop stack frame
+	add sp, sp, 0xa0 + TEB_OFFSET
+
+	; Jump to return address (in x30)
+	ret
+
+	endp
+
+	end
diff --git a/coroutine/arm64/Context.h b/coroutine/arm64/Context.h
index 1472621f..fe437e2c 100644
--- a/coroutine/arm64/Context.h
+++ b/coroutine/arm64/Context.h
@@ -15,7 +15,13 @@
 #include <stdint.h>
 #include <string.h>
 
+#if defined __GNUC__
 #define COROUTINE __attribute__((noreturn)) void
+#define COROUTINE_DECL COROUTINE
+#elif defined _MSC_VER
+#define COROUTINE __declspec(noreturn) void
+#define COROUTINE_DECL void
+#endif
 
 enum {COROUTINE_REGISTERS = 0xb0 / 8};
 
@@ -44,7 +50,7 @@ struct coroutine_context
 #endif
 };
 
-typedef COROUTINE(* coroutine_start)(struct coroutine_context *from, struct coroutine_context *self);
+typedef COROUTINE_DECL(* coroutine_start)(struct coroutine_context *from, struct coroutine_context *self);
 
 static inline void coroutine_initialize_main(struct coroutine_context * context) {
     context->stack_pointer = NULL;
diff --git a/include/ruby/atomic.h b/include/ruby/atomic.h
index 3eb80fbf..d4b7023f 100644
--- a/include/ruby/atomic.h
+++ b/include/ruby/atomic.h
@@ -404,7 +404,7 @@ rbimpl_atomic_size_add(volatile size_t *ptr, size_t val)
 #elif defined(HAVE_GCC_SYNC_BUILTINS)
     __sync_add_and_fetch(ptr, val);
 
-#elif defined(_WIN32) && defined(_M_AMD64)
+#elif defined(_WIN64)
     /* Ditto for `InterlockeExchangedAdd`. */
     InterlockedExchangeAdd64(ptr, val);
 
@@ -456,7 +456,7 @@ rbimpl_atomic_size_inc(volatile size_t *ptr)
 #elif defined(HAVE_GCC_ATOMIC_BUILTINS) || defined(HAVE_GCC_SYNC_BUILTINS)
     rbimpl_atomic_size_add(ptr, 1);
 
-#elif defined(_WIN32) && defined(_M_AMD64)
+#elif defined(_WIN64)
     InterlockedIncrement64(ptr);
 
 #elif defined(__sun) && defined(HAVE_ATOMIC_H) && (defined(_LP64) || defined(_I32LPx))
@@ -538,7 +538,7 @@ rbimpl_atomic_size_sub(volatile size_t *ptr, size_t val)
 #elif defined(HAVE_GCC_SYNC_BUILTINS)
     __sync_sub_and_fetch(ptr, val);
 
-#elif defined(_WIN32) && defined(_M_AMD64)
+#elif defined(_WIN64)
     const ssize_t neg = -1;
     InterlockedExchangeAdd64(ptr, neg * val);
 
@@ -590,7 +590,7 @@ rbimpl_atomic_size_dec(volatile size_t *ptr)
 #elif defined(HAVE_GCC_ATOMIC_BUILTINS) || defined(HAVE_GCC_SYNC_BUILTINS)
     rbimpl_atomic_size_sub(ptr, 1);
 
-#elif defined(_WIN32) && defined(_M_AMD64)
+#elif defined(_WIN64)
     InterlockedDecrement64(ptr);
 
 #elif defined(__sun) && defined(HAVE_ATOMIC_H) && (defined(_LP64) || defined(_I32LPx))
@@ -688,7 +688,7 @@ rbimpl_atomic_size_exchange(volatile size_t *ptr, size_t val)
 #elif defined(HAVE_GCC_SYNC_BUILTINS)
     return __sync_lock_test_and_set(ptr, val);
 
-#elif defined(_WIN32) && defined(_M_AMD64)
+#elif defined(_WIN64)
     return InterlockedExchange64(ptr, val);
 
 #elif defined(__sun) && defined(HAVE_ATOMIC_H) && (defined(_LP64) || defined(_I32LPx))
@@ -823,7 +823,7 @@ rbimpl_atomic_size_cas(volatile size_t *ptr, size_t oldval, size_t newval)
 #elif defined(HAVE_GCC_SYNC_BUILTINS)
     return __sync_val_compare_and_swap(ptr, oldval, newval);
 
-#elif defined(_WIN32) && defined(_M_AMD64)
+#elif defined(_WIN64)
     return InterlockedCompareExchange64(ptr, newval, oldval);
 
 #elif defined(__sun) && defined(HAVE_ATOMIC_H) && (defined(_LP64) || defined(_I32LPx))
diff --git a/include/ruby/internal/memory.h b/include/ruby/internal/memory.h
index 6884db19..e1c993ec 100644
--- a/include/ruby/internal/memory.h
+++ b/include/ruby/internal/memory.h
@@ -40,7 +40,12 @@
 
 #if defined(_MSC_VER) && defined(_WIN64)
 # include <intrin.h>
+# if defined(_M_AMD64)
 # pragma intrinsic(_umul128)
+# endif
+# if defined(_M_ARM64)
+# pragma intrinsic(__umulh)
+# endif
 #endif
 
 #include "ruby/internal/attr/alloc_size.h"
@@ -577,12 +582,16 @@ rbimpl_size_mul_overflow(size_t x, size_t y)
     ret.left  = dz > SIZE_MAX;
     ret.right = RBIMPL_CAST((size_t)dz);
 
-#elif defined(_MSC_VER) && defined(_WIN64)
+#elif defined(_MSC_VER) && defined(_M_AMD64)
     unsigned __int64 dp = 0;
     unsigned __int64 dz = _umul128(x, y, &dp);
     ret.left  = RBIMPL_CAST((bool)dp);
     ret.right = RBIMPL_CAST((size_t)dz);
 
+#elif defined(_MSC_VER) && defined(_M_ARM64)
+    ret.left  = __umulh(x, y) != 0;
+    ret.right = x * y;
+
 #else
     /* https://wiki.sei.cmu.edu/confluence/display/c/INT30-C.+Ensure+that+unsigned+integer+operations+do+not+wrap */
     ret.left  = (y != 0) && (x > SIZE_MAX / y);
diff --git a/vm_dump.c b/vm_dump.c
index e31d70c0..0611e3c9 100644
--- a/vm_dump.c
+++ b/vm_dump.c
@@ -727,6 +727,14 @@ dump_thread(void *arg)
                     frame.AddrFrame.Offset = context.Rbp;
                     frame.AddrStack.Mode = AddrModeFlat;
                     frame.AddrStack.Offset = context.Rsp;
+#elif defined(_M_ARM64) || defined(__aarch64__)
+                    mac = IMAGE_FILE_MACHINE_ARM64;
+                    frame.AddrPC.Mode = AddrModeFlat;
+                    frame.AddrPC.Offset = context.Pc;
+                    frame.AddrFrame.Mode = AddrModeFlat;
+                    frame.AddrFrame.Offset = context.Fp;
+                    frame.AddrStack.Mode = AddrModeFlat;
+                    frame.AddrStack.Offset = context.Sp;
 #else	/* i386 */
                     mac = IMAGE_FILE_MACHINE_I386;
                     frame.AddrPC.Mode = AddrModeFlat;
diff --git a/win32/Makefile.sub b/win32/Makefile.sub
index b8c16b8b..d4a84027 100644
--- a/win32/Makefile.sub
+++ b/win32/Makefile.sub
@@ -331,6 +331,9 @@ COROUTINE_SRC = $(COROUTINE_OBJ:.obj=.asm)
 !elseif "$(ARCH)" == "i386"
 COROUTINE_OBJ = coroutine/win32/Context.obj
 COROUTINE_SRC = $(COROUTINE_OBJ:.obj=.asm)
+!elseif "$(ARCH)" == "arm64"
+COROUTINE_OBJ = coroutine/arm64/Context.obj
+COROUTINE_SRC = $(COROUTINE_OBJ:.obj=.asm)
 !else
 COROUTINE_OBJ = coroutine/copy/Context.obj
 COROUTINE_SRC = $(COROUTINE_OBJ:.obj=.c)
@@ -678,7 +681,7 @@ $(CONFIG_H): $(MKFILES) $(srcdir)/win32/Makefile.sub $(win_srcdir)/Makefile.sub
 #define _INTEGRAL_MAX_BITS 64
 #endif
 #define SIZEOF_OFF_T 8
-!if "$(ARCH)" == "x64"
+!if "$(TARGET_OS)" == "mswin64"
 #define SIZEOF_VOIDP 8
 !else
 #define SIZEOF_VOIDP 4
@@ -700,7 +703,7 @@ $(CONFIG_H): $(MKFILES) $(srcdir)/win32/Makefile.sub $(win_srcdir)/Makefile.sub
 #define NUM2CLOCKID(v) NUM2INT(v)
 #define SIZEOF_CLOCK_T 4
 #define SIZEOF_RLIM_T 0
-!if "$(ARCH)" == "x64"
+!if "$(TARGET_OS)" == "mswin64"
 #define SIZEOF_SIZE_T 8
 #define SIZEOF_PTRDIFF_T 8
 #define SIZEOF_INTPTR_T 8
@@ -807,7 +810,7 @@ $(CONFIG_H): $(MKFILES) $(srcdir)/win32/Makefile.sub $(win_srcdir)/Makefile.sub
 #define HAVE_INTPTR_T 1
 #define HAVE_UINTPTR_T 1
 #define HAVE_SSIZE_T 1
-!if "$(ARCH)" == "x64"
+!if "$(TARGET_OS)" == "mswin64"
 #define ssize_t __int64
 #define PRI_PTR_PREFIX "I64"
 !else
@@ -1286,6 +1289,9 @@ $(ruby_pc): $(RBCONFIG)
 {$(srcdir)/coroutine/win64}.asm{coroutine/win64}.obj:
 	$(ECHO) assembling $(<:\=/)
 	$(Q) $(AS) $(ASFLAGS) $(XCFLAGS) $(CPPFLAGS) $(COUTFLAG)$@ -c $(<:\=/)
+{$(srcdir)/coroutine/arm64}.asm{coroutine/arm64}.obj:
+	$(ECHO) assembling $(<:\=/)
+	$(Q) $(AS) $(ASFLAGS) -o $@ $(<:\=/)
 
 {$(srcdir)/enc/trans}.c.obj:
 	$(ECHO) compiling $(<:\=/)
diff --git a/win32/setup.mak b/win32/setup.mak
index a1a0ce49..b8e61349 100644
--- a/win32/setup.mak
+++ b/win32/setup.mak
@@ -227,7 +227,9 @@ RUBY_SO_NAME = $(RUBY_SO_NAME)
 
 -generic-: nul
 	@$(CPP) <<conftest.c 2>nul | findstr = >>$(MAKEFILE)
-#if defined _M_X64
+#if defined _M_ARM64
+MACHINE = arm64
+#elif defined _M_X64
 MACHINE = x64
 #else
 MACHINE = x86
@@ -241,6 +243,8 @@ MACHINE = x86
 	@echo MACHINE = alpha>>$(MAKEFILE)
 -x64-: -osname64-
 	@echo MACHINE = x64>>$(MAKEFILE)
+-arm64-: -osname64-
+	@echo MACHINE = arm64>>$(MAKEFILE)
 -ix86-: -osname32-
 	@echo MACHINE = x86>>$(MAKEFILE)
 
-- 
2.50.1 (Apple Git-155)

